feat(search): introduce targeted search

asudoh opened the pull request:
Fixes #3513.

#### Changelog

**New**

- The "targeted search" component.

#### Testing / Reviewing

The new component can be found at Search - targeted story.

asudoh committed:
feat(search): introduce targeted search

Fixes #3513.

asudoh committed:
chore(build): add an export

asudoh committed:
test(exports): update exports

emyarod reviewed (commented):
it looks like if a dropdown option is selected, it is removed from the dropdown menu, but the original 
screenshot in #3513 still shows the option in the dropdown menu. how should the dropdown behave here?

also it looks like the borders around the entire component are more in line with v9 spec right?

asudoh commented:
@emyarod Good points - Wrt selection will incorporate @elizabethsjudd's change for that. Wrt border will leave it for our designers.

elizabethsjudd commented:
When using the dropdown component you can now add the `bx--dropdown--show-selected` modifier and have the selected item show up in the list. Here's the PR where I made that change: https://github.com/carbon-design-system/carbon/pull/4406

asudoh committed:
fix(search): show selected dropdown item

asudoh commented:
Thanks @elizabethsjudd - Updated with visible selected item.

emyarod reviewed (commented):
on vanilla there is an extra tab stop in between the dropdown and the text input when the dropdown is closed

more of a question for design, but do we want the borders on all 4 sides of the component or just the bottom border? cc @designertyler 

on a side note (this is probably for a separate issue) we should align the focus handling in the React and vanilla dropdowns/listbox

asudoh commented:
Thanks @emyarod for sharing your thoughts!

> on vanilla there is an extra tab stop in between the dropdown and the text input when the dropdown is closed

Seems that it's an issue with vanilla dropdown. Reported at: #4640

> more of a question for design, but do we want the borders on all 4 sides of the component or just the bottom border? cc @designertyler

I assumed the four-sides borders was the design intent, but good to have designer's response here: CC @aagonzales 

> on a side note (this is probably for a separate issue) we should align the focus handling in the React and vanilla dropdowns/listbox

Yes, from my point of view it's another separate topic of dropdown.

designertyler commented:
> do we want the borders on all 4 sides of the component or just the bottom border

We should only have the border on the bottom so it's consistent with the rest of our fields.
![Screen Shot 2019-11-11 at 8 50 09 AM](https://user-images.githubusercontent.com/19270502/68596175-6c62b080-0460-11ea-9a2f-bd6151c8ced6.png)



The border between the dropdown and the search field should be a gap similar to the data range picker 
![Screen Shot 2019-11-11 at 8 48 53 AM](https://user-images.githubusercontent.com/19270502/68596056-33c2d700-0460-11ea-9dc6-a82966cbfdb4.png)



The expected specs should be like this:
![image](https://user-images.githubusercontent.com/19270502/68596416-ccf1ed80-0460-11ea-9f3a-ccb7b8a04b1f.png)

asudoh committed:
fix(search): fixes upon review comments

asudoh commented:
Thanks @designertyler for reviewing! Fixed.

emyarod commented on packages/components/src/components/search/_search.scss:
```diff
@@ -240,6 +240,84 @@
+
+    .#{$prefix}--search--xl .#{$prefix}--search-input {
+      padding-left: rem(14px);
+      padding-right: rem(40px);
```
just wondering, how are we deriving some of these values? typically I'm expecting spacing tokens or values equal to our spacing tokens but not sure if these are exceptions based on a spec. same with L300-301 and L313-314

asudoh replied:
Good questions @emyarod, these are for adjusting the spacings to different sizes of search box. If designers have specific spacing tokens to use, we can make changes to use them. CC @designertyler @aagonzales

aagonzales replied:
I don't think either of those values are correct. 
padding-left: 16px / `$spacing-05`
padding-right: should be 48px so the icons is centered in an invisible square.

![image](https://user-images.githubusercontent.com/11670886/68810608-38d87f80-0634-11ea-927c-2276946c5db6.png)

asudoh replied:
Thanks @aagonzales for clarifying - A follow-up question; What should happen with medium/small size variants? Or targeted search should have only the default size variant?

aagonzales replied:
it would stay centered in a squared container like this:

![image](https://user-images.githubusercontent.com/11670886/68889465-9a572780-06e2-11ea-8320-45e40b7addc9.png)

vpicone commented:
I'm wondering why this is a component now as opposed to just a pattern like @designertyler recommended in #3513. 

This is such a simple combination of components, do we plan on making 'wrappers' for other combinations of components/patterns? It seems like it could get out of hand.

Sidenote: also wondering why the container takes it's own size prop, shouldn't it just let the internal components set the size? This means the developer will have to pass in the same value to two locations.

asudoh committed:
Merge branch 'master' into search-targeted

asudoh committed:
fix(search): update magnifier size and left spacing

asudoh commented:
Fixed paddings. Changed to use spacing token for left padding. Right padding doesn't still use spacing token given it's based on magnifier size.

@aagonzales Please let me know if you think magnifier size is based on spacing token. BTW it appears that dropdown changing its size (upon search size) is specific to targeted search rather than it's for wider use, is it correct? Similar question to search box with magnifier at right and without close button, which also seems specific to target search UI. Thanks!

aagonzales commented:
- Yes, the dropdown height and search field height should always be paired in this scenario. 
- Also looks like we're not following the [original spec](https://github.com/carbon-design-system/carbon/issues/3513#issue-471830542), I probably should have looked at that before jumping in. But we have the search icon on the wrong side. It should be on the left side to match with the normal search as well as allow for the "clear" x to appear on the right side. There are some adjust's to the padding spec I previous shared. Sorry about that! 

![image](https://user-images.githubusercontent.com/11670886/68964115-2aa67280-079e-11ea-909b-93de15838a3f.png)

------

> BTW it appears that dropdown changing its size (upon search size) is specific to targeted search rather than it's for wider use, is it correct? 

- Option 2 here would be the ideal scenario, with the field and menu width decoupled. I think its pretty important to see this full string of text since it's the search parameters. Its primary content and shouldn't be truncated if it can be helped. 

![image](https://user-images.githubusercontent.com/11670886/68964083-15314880-079e-11ea-9ebf-a3db1146fbe0.png)

Amazon's  targeted search does this (though it's using a select)
![Nov-15-2019 10-35-42](https://user-images.githubusercontent.com/11670886/68964277-8ffa6380-079e-11ea-9a86-95df01db532e.gif)

asudoh commented:
>> BTW it appears that dropdown changing its size (upon search size) is specific to targeted search rather than it's for wider use, is it correct?

> Option 2 here would be the ideal scenario, with the field and menu width decoupled. I think its pretty important to see this full string of text since it's the search parameters. Its primary content and shouldn't be truncated if it can be helped.

Thanks @aagonzales for some clarifications! Any thoughts on the vertical size of the dropdown? I'm assuming that the dropdown's vertical size should change upon the choice of the small/large/extra-large search box next to the dropdown, and such (vertical) size choice of dropdown shouldn't happen with regular usage of dropdown (dropdown without a search box). Is it correct?

asudoh committed:
Merge remote-tracking branch 'upstream/master' into search-targeted

asudoh committed:
fix(search): fix magnifier position and close button

asudoh commented:
Fixed the magnifier and the close button. Made a change to yield constant width regardless of the dropdown content. @aagonzales Would like to have your thoughts on the dropdown height still - To make the long story short, the question is whether non-40px-height dropdown should be used for non-search use case or not. Thanks!

aagonzales commented:
@asudoh yes, we want dropdown in other sizes for general purposes, it won't be just for search. (Funnily enough I just got a question today asking about this topic).

joshblack commented:
@dakahn do we have a test procedure for this on the a11y side of things? Would it be helpful to include IBMa to run tests here if not?

designertyler commented:
What's the status of this PR? Based on Anna's design specs it's pretty close. Only issue I still see is there needs to be a gap between the dropdown and search field. Right now they're touching

![image](https://user-images.githubusercontent.com/19270502/70746110-ea79d780-1cea-11ea-8513-02b57ddfc1cf.png)

alisonjoseph commented:
@designtyler this/new feature work is on hold for now until we have the time to fully test accessibility implications. We also need designer bandwidth to update the sketch kit and update documentation on the website around this new component (or is it a pattern?)

joshblack commented:
Also seems like @vpicone had some good questions/points over in https://github.com/carbon-design-system/carbon/pull/4583#issuecomment-553700723 that should also be addressed

dakahn commented:
Assigning myself to do an Accessibility Verification Test pass on this PR 👍 @joshblack

dakahn reviewed (changes requested):
#### Tested on Windows 10 with Firefox latest using NVDA

1. When focus is on the document — the next tabstop is the text input and not the listbox. At a glance it seems like we have `tabindex="-1"` on the div with `id="dropdown-2"` in the listbox — that might be Downshift and not us, but i'd expect the listbox to be first in the tab order (since that's how it appears visually and should be logically).

2. The targeted search list describes itself as "list clickable open menu menu button collapse submenu clickable". It looks like we have an `aria-label` of "open menu" on the outermost div. That's an inadequate description. I'd expect something along the lines of "Targeted search options. Option 1 selected." — Clearly letting the user know A) what this listbox does and B) that we have a default narrowing of their search.

3. When the options list is opened via the keyboard it tells the user how many options are available and to "use up and down arrow keys to navigate", but when the user hits down arrow the list box closes and focus is lost (returned to the document). (will follow up with VoiceOver and JAWS testing).

4. The input box for the search is described as "Search edit has autocomplete. Option 1". If our intention is to have the user understand the scope of their search is limited to the selection of the box on the left we should be a little more descriptive. I'd expect something along the lines of "Targeted search has autocomplete. Option 1 selected."

#### Tested on macOS latest with VoiceOver on Safari

1. First tabstop skips listbox as well (see 1 above)

2. The targeted search list describes itself much less verbosely, but is still inadequately labeled (see 2 above).

3. The options list is keyboard accessible using the arrow keys, but since Option 1 is selected by default it must be focused when the listbox is expanded (right now it's merely selected). 

4. See 3 above

Some of these might be problems with the Listbox component in general. In that case we can roll them out into issue tickets and not have them block this PR 👍

tw15egan committed:
Merge branch 'master' into search-targeted

joshblack commented:
Hey there! Going through the PR backlog today and am clearing out PRs that have been up for a while 👀
